File: typescript/packages/core/src/core/qdrant/payload.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { QdrantRow } from './client.ts'

/** Read a Qdrant payload field, including `metadata.source`-style nested keys. */
export function fieldValue(row: QdrantRow, field: string | null): unknown {
  if (field === null || field === '') return undefined
  if (!field.includes('.')) return row[field]
  let value: unknown = row
  for (const part of field.split('.')) {
    if (typeof value !== 'object' || value === null || Array.isArray(value)) return undefined
    const record = value as Record<string, unknown>
    if (!Object.hasOwn(record, part)) return undefined
    value = record[part]
  }
  return value
}

/** Copy a payload while removing one dotted field path. */
export function withoutField(row: QdrantRow, field: string | null): QdrantRow {
  if (field === null || field === '') return { ...row }
  return omitPath(row, field.split('.'))
}

function omitPath(row: QdrantRow, parts: string[]): QdrantRow {
  const [head, ...tail] = parts
  const copied: QdrantRow = {}
  for (const [key, value] of Object.entries(row)) {
    if (key !== head) {
      copied[key] = value
    } else if (tail.length > 0) {
      copied[key] =
        typeof value === 'object' && value !== null && !Array.isArray(value)
          ? omitPath(value as QdrantRow, tail)
          : value
    }
  }
  return copied
}

File: python/mirage/core/qdrant/payload.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from collections.abc import Mapping
from typing import Any


def field_value(row: Mapping[str, Any], field: str | None) -> Any:
    """Read a Qdrant payload field, including dotted nested keys.

    Qdrant spells nested payload paths with dots in filters. Mirroring that
    spelling in config means ``metadata.source`` addresses
    ``{"metadata": {"source": ...}}`` everywhere the mount reads a field.

    Args:
        row (Mapping[str, Any]): point payload plus Mirage's synthetic fields.
        field (str | None): configured payload path.
    """
    if not field:
        return None
    if "." not in field:
        return row.get(field)
    value: Any = row
    for part in field.split("."):
        if not isinstance(value, Mapping) or part not in value:
            return None
        value = value[part]
    return value


def without_field(row: Mapping[str, Any], field: str | None) -> dict[str, Any]:
    """Copy a payload while removing one dotted field path.

    Args:
        row (Mapping[str, Any]): point payload plus Mirage's synthetic fields.
        field (str | None): configured payload path to drop.
    """
    copied = dict(row)
    if not field:
        return copied
    parts = field.split(".")
    if len(parts) == 1:
        copied.pop(field, None)
        return copied
    source: Any = row
    target: dict[str, Any] = copied
    for part in parts[:-1]:
        if not isinstance(source, Mapping):
            return copied
        child = source.get(part)
        if not isinstance(child, Mapping):
            return copied
        cloned = dict(child)
        target[part] = cloned
        source = child
        target = cloned
    target.pop(parts[-1], None)
    return copied
